## The “Day One” challenge from the 2015 Advent of Code

All Advent of Code challenges have two parts, with the second part building on the first. We tackled both parts of the Day One challenge of the 2015 edition.

### Part one

#### The challenge

Here’s the challenge, taken straight from [its webpage](https://adventofcode.com/2015/day/1):

> Santa is trying to deliver presents in a large apartment building, but he can't find the right floor - the directions he got are a little confusing. He starts on the ground floor (floor 0) and then follows the instructions one character at a time.
>
> An opening parenthesis, `(`, means he should go *up* one floor, and a closing parenthesis, `)`, means he should go *down* one floor.
>
> The apartment building is very tall, and the basement is very deep; he will never find the top or bottom floors.
>
> For example:
>
> - `(())` and `()()` both result in floor 0.
> - `(((` and `(()(()(` both result in floor 3.
> - `))(((((` also results in floor 3.
> - `())` and `))(` both result in floor -1 (the first basement level).
> - `)))` and `)())())` both result in floor -3.
> 
> To what floor do the instructions take Santa?

Every participant in Advent of Code gets a different input for their solution. This was the input given to me:

```
()()(()()()(()()((()((()))((()((((()()((((()))()((((())(((((((()(((((((((()(((())(()()(()((()()(()(())(()((((()((()()()((((())((((((()(()(((()())(()((((()))())(())(()(()()))))))))((((((((((((()())()())())(())))(((()()()((((()(((()(()(()()(()(()()(()(((((((())(())(())())))((()())()((((()()((()))(((()()()())))(())))((((())(((()())(())(()))(()((((()())))())((()(())(((()((((()((()(())())))((()))()()(()(()))))((((((((()())((((()()((((()(()())(((((()(()())()))())(((()))()(()(()(()((((()(())(()))(((((()()(()()()(()(((())())(((()()(()()))(((()()(((())())(()(())())()()(())()()()((()(((()(())((()()((())()))((()()))((()()())((((()(()()(()(((()))()(()))))((()(((()()()))(()(((())()(()((()())(()(()()(()())(())()(((()(()())()((((()((()))))())()))((()()()()(())()())()()()((((()))))(()(((()()(((((((())()))()((((()((())()(()())(())()))(()(()())(((((((())))(((()))())))))()))())((())(()()((())()())()))))()((()()())(())((())((((()())())()()()(((()))())))()()))())(()()()(()((((((()()))())()))()(((()(((())((((()()()(()))())()()))))())()))())((())()())(((((())())((())())))(((())(((())(((((()(((((())(()(()())())(()(())(()))(()((((()))())()))))())))((()(()))))())))(((((())()))())()))))()))))(((()))()))))((()))((()((()(()(())()())))(()()()(())()))()((((())))))))(())(()((()()))(()))(()))(()((()))))))()()((((()()))()())()))))))()()()))(()((())(()))((()()()())()(((()((((())())))()((((()(()))))))())))()()())()))(()))))(()())()))))))((())))))))())()))()((())())))(()((()))()))(())))))(()))()())()()))((()(()))()()()()))))())()()))())(())()()))()))((()))))()()(()())))))()()()))((((()))()))))(()(())))(()())))((())())(()))()))))()())))()())()())))))))))()()))))())))((())((()))))())))(((()())))))))(()))()()))(()))()))))()())))))())((((()())))))))())))()()))))))))()))()))))()))))))(())))))))))())))))))))))))))())())((())))))))))()))((())))()))))))))())()(()))))))())))))()()()())()(()()()(()())(()))()()()(()())))())())))()))))())))))))()()()()())(())())()())()))))(()()()()()))))()))())())))((()())()())))()))()))))(()())))()))))))))(((()))()()))))))))))))))))))))(()))(()((()))())))())(()))(()(()(())))))()(()))()))()()))))))))))))()((()())(())())()(())))))())()())((()()))))(()()))))())()(())()))))))))))))))))))))()))(()(()())))))))()()((()))()))))))((())))()))))))))((()))())()()))())()()))((()))())))))))))))(()())()))(())((()(()()))(()())(())))()())(()(())()()))))()))()(()))))))(()))))))))))(()))())))))))))())))))())))(())))))()))))(())())))))))))()(()))))()())))())(()))()())))))))))))))())()()))))()))))))())))))()))))(())(()()()()((())()))())(()))((())()))())())(())(()()))))()))(())()()((())(())))(())))()))())))))))))()(((((())())))(())()))))(())))((()))()(((((((()))))()()))(())))))()(()))))(()()))()))())))))))(()())()))))))))())))(()))())()))(())()((())())()())())(()(()))))()))))))((()())(())()()(()())))()()))(())(())(()))())))()))(()))()()))((((()))))()))((()()()))))()))()))())))(()))()))))(())))()))())()(()))()())))())))))))())))())))()()))))))(()))())())))()))()()())())))))))))))))())))()))(()()))))())))())()(())))())))))))))))))))))()()())())))))()()()((()(()))()()(())()())()))()))))()()()))))))((()))))))))()(()(()((((((()()((()())))))))))))()))())))))((())())(()))())))())))))())()()())(())))())))()())())(())))))))()()(())))()))())))())())())()))))))))()))(()()()())())())))(())())))))))()()())()))))())))())()(())())))))))()())()))(()()(())())))()(()((()()((()()(((((())(()())()))(())()))(())))(())))))))()))()))((()))()))()))))))))()))))))))((()()())(()))(((()))(())))()))((())(((())))()())))())))))((())))))(())())((((((())())()(()))()(()((()())))((())()(()(()))))(())(()()())(())))())((()(((())())))(((()())())))())()(())())((((()()))))())((()))()()()()(())(((((((()()()((()))())(()())))(())())((((()()(()))))()((())))((())()))()(((()))())))()))((()(()))(())(()((((())((((()()(()()))(((())(()))))((((()(()))(())))))((()))(()))((()(((()(()))(()(()((()(())(()(()(()(()()((()))())(((())(()(()))))(()))()()))(())))(())()(((())(()))()((((()()))))())(()))))((())()((((()(((()))())())(((()))()())((())(())())(())()(())()(()()((((((()()))))()()(((()()))))()())()(((()(()))(()(()())(()(()))))(((((()(((())())))))(((((()((()()((())())((((((()(())(()()((()()()()()()()(()()))()(((()))()))(((((((())(((()((()())()((((())(((()(())))()((()(()()()((())((()())()))()))())))())((((((()))(()(()()()))(()((()(()(()))()((()(((()()()((())(((((())()(()))())())((()(())))(()(()())(())((())())())(((()()()(())))))())(()))))))()))))))())((()()()))((()((((((()))(((()((((()()()(((()))())()(()()(((()((()()()()())()()))()()()(()(())((()))))(()))())))))))()(()()(((((())()(()(((((()((()(()()())(()((((((((()((((((())()((((()()()((()((()((((((()))((())))))))())()))((()(()))()(()()(()((())((()()((((((((((((()())(()()()))((((()((((((())(()))())(()()((()()))()(((((((()((()()((((((()(((())))((())))((((((((()()(((((((())(((((()())(((())((())()((((()(((((((()(()(((()((((((()(((()(((((((((((()()((()()(()))((()()(((()(((())))((((())()(()(((())()(()(((())(((((((((((()))())))((((((())((()()((((()())())((((()()))((())(((((()(()()(()()()((())(()((()()((((()(((((()((()(()((((()())((((((()(((((()()(()(()((((())))(())(())(())((((()(()()((((()((((()()((()((((((())))(((((()))))()))(()((((((((()(((())())(((())))(()(()((())(((()((()()(((((()((()()(((())()(()))(((((((())(()(((((()))((()((()((()))(())())((((()((((())()(()))(((()(((((((((((((((())(((((((((()))(((()(()()()()((((((()((())()((((((((()(())(((((((((((()(()((())()((()()(()(()()((((()()((())(()((()()(()()((((()(((((((())))((((())(())()(((()()((()()((((()((()(((()((())(((()()()((((()((((()()(()(()((((((((())(()(((((())(()())(((((((()())()(()((((()((())(()()())((((()()(((()((((())(())(()()(((((((((()()))()(((())(()(()((((((())(()()())(()))()()(((()(((()((())(()(((((((()(()(()((()(((((()(()((()(()((((((()((((()()((((()(((()((())(()(()((()()((((()()(())()(())(((())(()((((((((()())(((((((((()(())()((((())))()))()()(((((()()((((((())(()()(((()(()(((((((()(()(((((((())(())((((()((()(())))((((()()())(()))((()())((((()(((((()(()(())(()(()()())(((((()(((((()((((()()((((((((()()))(()((((((())((((())()(()(((()()()(((()(()(())(())(((((()(())())((((())(())(()(((()(((((())((((())())((()(((((((()(((())(()(()))(((((((((()((()((()()(()((((())(((()((())((((())(()(((()(((()(()((((()(((())(()(((()(()()(()(()((()()(()())(())())((()(()(((()(((()(((()()(((((((((()(((((((((()()(((()(((()())((((()(()(((()()()((())((((((((((())(()(((()((((()())((((()((()))(((()()()(((((()(((((((())((()())(()((((())((((((((())(()((()((((((((((()()((()((()()))(((()())()())()(((()())()()(()(()(((((((())()))(())()))())()()((())()((()((((()((()((())(((((()((((((()(())))(()))())(((()))((()()(()(((()))((((())()(((()))))()(()(())()(((((())(()(()(())(())()((()()()((((()(())((()())(()(()))(()(()(()()(())()()(()((())()((()))))()))((()(()()()()((()())(()))())()(()(((((((((())())((()((()((((((())()((((())(((())((()(()()()((())(()((())(((()((((()()((()(()(((((())()))()((((((()))((())(((()()))(((())(())()))(((((((())(())())()(())(((((()))()((()))()(()()((()()()()()())(((((((
```

#### Importing the input into Python

The simplest way to bring the input into Python is to copy the input, paste it into a triple-quoted string, and assign the string to a variable, which I named `raw_input`:

In [78]:
raw_input = """()()(()()()(()()((()((()))((()((((()()((((()))()((((())(((((((()(((((((((()(((())(()()(()((()()(()(())(()((((()((()()()((((())((((((()(()(((()())(()((((()))())(())(()(()()))))))))((((((((((((()())()())())(())))(((()()()((((()(((()(()(()()(()(()()(()(((((((())(())(())())))((()())()((((()()((()))(((()()()())))(())))((((())(((()())(())(()))(()((((()())))())((()(())(((()((((()((()(())())))((()))()()(()(()))))((((((((()())((((()()((((()(()())(((((()(()())()))())(((()))()(()(()(()((((()(())(()))(((((()()(()()()(()(((())())(((()()(()()))(((()()(((())())(()(())())()()(())()()()((()(((()(())((()()((())()))((()()))((()()())((((()(()()(()(((()))()(()))))((()(((()()()))(()(((())()(()((()())(()(()()(()())(())()(((()(()())()((((()((()))))())()))((()()()()(())()())()()()((((()))))(()(((()()(((((((())()))()((((()((())()(()())(())()))(()(()())(((((((())))(((()))())))))()))())((())(()()((())()())()))))()((()()())(())((())((((()())())()()()(((()))())))()()))())(()()()(()((((((()()))())()))()(((()(((())((((()()()(()))())()()))))())()))())((())()())(((((())())((())())))(((())(((())(((((()(((((())(()(()())())(()(())(()))(()((((()))())()))))())))((()(()))))())))(((((())()))())()))))()))))(((()))()))))((()))((()((()(()(())()())))(()()()(())()))()((((())))))))(())(()((()()))(()))(()))(()((()))))))()()((((()()))()())()))))))()()()))(()((())(()))((()()()())()(((()((((())())))()((((()(()))))))())))()()())()))(()))))(()())()))))))((())))))))())()))()((())())))(()((()))()))(())))))(()))()())()()))((()(()))()()()()))))())()()))())(())()()))()))((()))))()()(()())))))()()()))((((()))()))))(()(())))(()())))((())())(()))()))))()())))()())()())))))))))()()))))())))((())((()))))())))(((()())))))))(()))()()))(()))()))))()())))))())((((()())))))))())))()()))))))))()))()))))()))))))(())))))))))())))))))))))))))())())((())))))))))()))((())))()))))))))())()(()))))))())))))()()()())()(()()()(()())(()))()()()(()())))())())))()))))())))))))()()()()())(())())()())()))))(()()()()()))))()))())())))((()())()())))()))()))))(()())))()))))))))(((()))()()))))))))))))))))))))(()))(()((()))())))())(()))(()(()(())))))()(()))()))()()))))))))))))()((()())(())())()(())))))())()())((()()))))(()()))))())()(())()))))))))))))))))))))()))(()(()())))))))()()((()))()))))))((())))()))))))))((()))())()()))())()()))((()))())))))))))))(()())()))(())((()(()()))(()())(())))()())(()(())()()))))()))()(()))))))(()))))))))))(()))())))))))))())))))())))(())))))()))))(())())))))))))()(()))))()())))())(()))()())))))))))))))())()()))))()))))))())))))()))))(())(()()()()((())()))())(()))((())()))())())(())(()()))))()))(())()()((())(())))(())))()))())))))))))()(((((())())))(())()))))(())))((()))()(((((((()))))()()))(())))))()(()))))(()()))()))())))))))(()())()))))))))())))(()))())()))(())()((())())()())())(()(()))))()))))))((()())(())()()(()())))()()))(())(())(()))())))()))(()))()()))((((()))))()))((()()()))))()))()))())))(()))()))))(())))()))())()(()))()())))())))))))())))())))()()))))))(()))())())))()))()()())())))))))))))))())))()))(()()))))())))())()(())))())))))))))))))))))()()())())))))()()()((()(()))()()(())()())()))()))))()()()))))))((()))))))))()(()(()((((((()()((()())))))))))))()))())))))((())())(()))())))())))))())()()())(())))())))()())())(())))))))()()(())))()))())))())())())()))))))))()))(()()()())())())))(())())))))))()()())()))))())))())()(())())))))))()())()))(()()(())())))()(()((()()((()()(((((())(()())()))(())()))(())))(())))))))()))()))((()))()))()))))))))()))))))))((()()())(()))(((()))(())))()))((())(((())))()())))())))))((())))))(())())((((((())())()(()))()(()((()())))((())()(()(()))))(())(()()())(())))())((()(((())())))(((()())())))())()(())())((((()()))))())((()))()()()()(())(((((((()()()((()))())(()())))(())())((((()()(()))))()((())))((())()))()(((()))())))()))((()(()))(())(()((((())((((()()(()()))(((())(()))))((((()(()))(())))))((()))(()))((()(((()(()))(()(()((()(())(()(()(()(()()((()))())(((())(()(()))))(()))()()))(())))(())()(((())(()))()((((()()))))())(()))))((())()((((()(((()))())())(((()))()())((())(())())(())()(())()(()()((((((()()))))()()(((()()))))()())()(((()(()))(()(()())(()(()))))(((((()(((())())))))(((((()((()()((())())((((((()(())(()()((()()()()()()()(()()))()(((()))()))(((((((())(((()((()())()((((())(((()(())))()((()(()()()((())((()())()))()))())))())((((((()))(()(()()()))(()((()(()(()))()((()(((()()()((())(((((())()(()))())())((()(())))(()(()())(())((())())())(((()()()(())))))())(()))))))()))))))())((()()()))((()((((((()))(((()((((()()()(((()))())()(()()(((()((()()()()())()()))()()()(()(())((()))))(()))())))))))()(()()(((((())()(()(((((()((()(()()())(()((((((((()((((((())()((((()()()((()((()((((((()))((())))))))())()))((()(()))()(()()(()((())((()()((((((((((((()())(()()()))((((()((((((())(()))())(()()((()()))()(((((((()((()()((((((()(((())))((())))((((((((()()(((((((())(((((()())(((())((())()((((()(((((((()(()(((()((((((()(((()(((((((((((()()((()()(()))((()()(((()(((())))((((())()(()(((())()(()(((())(((((((((((()))())))((((((())((()()((((()())())((((()()))((())(((((()(()()(()()()((())(()((()()((((()(((((()((()(()((((()())((((((()(((((()()(()(()((((())))(())(())(())((((()(()()((((()((((()()((()((((((())))(((((()))))()))(()((((((((()(((())())(((())))(()(()((())(((()((()()(((((()((()()(((())()(()))(((((((())(()(((((()))((()((()((()))(())())((((()((((())()(()))(((()(((((((((((((((())(((((((((()))(((()(()()()()((((((()((())()((((((((()(())(((((((((((()(()((())()((()()(()(()()((((()()((())(()((()()(()()((((()(((((((())))((((())(())()(((()()((()()((((()((()(((()((())(((()()()((((()((((()()(()(()((((((((())(()(((((())(()())(((((((()())()(()((((()((())(()()())((((()()(((()((((())(())(()()(((((((((()()))()(((())(()(()((((((())(()()())(()))()()(((()(((()((())(()(((((((()(()(()((()(((((()(()((()(()((((((()((((()()((((()(((()((())(()(()((()()((((()()(())()(())(((())(()((((((((()())(((((((((()(())()((((())))()))()()(((((()()((((((())(()()(((()(()(((((((()(()(((((((())(())((((()((()(())))((((()()())(()))((()())((((()(((((()(()(())(()(()()())(((((()(((((()((((()()((((((((()()))(()((((((())((((())()(()(((()()()(((()(()(())(())(((((()(())())((((())(())(()(((()(((((())((((())())((()(((((((()(((())(()(()))(((((((((()((()((()()(()((((())(((()((())((((())(()(((()(((()(()((((()(((())(()(((()(()()(()(()((()()(()())(())())((()(()(((()(((()(((()()(((((((((()(((((((((()()(((()(((()())((((()(()(((()()()((())((((((((((())(()(((()((((()())((((()((()))(((()()()(((((()(((((((())((()())(()((((())((((((((())(()((()((((((((((()()((()((()()))(((()())()())()(((()())()()(()(()(((((((())()))(())()))())()()((())()((()((((()((()((())(((((()((((((()(())))(()))())(((()))((()()(()(((()))((((())()(((()))))()(()(())()(((((())(()(()(())(())()((()()()((((()(())((()())(()(()))(()(()(()()(())()()(()((())()((()))))()))((()(()()()()((()())(()))())()(()(((((((((())())((()((()((((((())()((((())(((())((()(()()()((())(()((())(((()((((()()((()(()(((((())()))()((((((()))((())(((()()))(((())(())()))(((((((())(())())()(())(((((()))()((()))()(()()((()()()()()())((((((("""

#### The strategy

The strategy for this challenge is simple: Get a count of the following:

* The “up” parentheses: `(`
* The “down” parentheses: `)`

The floor that Santa ends up on will be (*number of “up” parentheses*) - (*number of “down” parentheses*).

#### One possible solution: Using a `for` loop to count `(` and `)` characters

One possible solution is to use a `for` to iterate over each character in `raw_input`. Here’s a quick example:

In [79]:
for character in "(()()(()":
    print(character)

(
(
)
(
)
(
(
)


We can use a loop like the one above to step through each character in `raw_input`, keeping a count of the number of `(` and `)` characters along the way:

In [80]:
open_parentheses = 0
close_parentheses = 0

for character in raw_input:
    if character == "(":
        open_parentheses += 1
    elif character == ")":
        close_parentheses += 1
    else:
        print("Something's wrong with input.")
        print(f"Unexpected character: {character}")
        quit()
        
print(f"Santa arrived on this floor: {open_parentheses - close_parentheses}")

Santa arrived on this floor: 280


I could have written the `if` statement like this...

```
for character in raw_input:
    if character == "(":
        open_parentheses += 1
    else:
        close_parentheses += 1
```

...but I wanted to demonstrate a little “defensive programming” against bad input. In the code above, any character that *wasn’t* a `)` would have increased the “down” count by one. 

The approach I took makes it very clear that only a `(` character would increase the “up” count by one, and only a `)` character would increase the “down” count by one. Any other character is unexpected, and the `else` clause makes it very clear that there’s something wrong with the input.

The `quit()` function causes a Python program to stop immediately.

The result of **280** was correct, and we could proceed to the next part of the challenge.

#### Another possible solution: Using the `count()` method to count `(` and `)` characters

Another solution requires knowing of the existence of the `count()` method, which is built into all strings. Given a *substring* — that’s just fancy talk for a part of a string — `count()` will return the number of occurrences of that substring within the string.

This is one of those moments where *showing* works better than *telling*, so here are some examples:

In [81]:
"aaaaa".count("a")

5

In [82]:
test_word = "abcabcabc"
print(test_word.count("abc"))

3


In [83]:
print(test_word.count("abcabcabc"))

1


Note that `count()` doesn’t count overlapping instances of a substring:

In [84]:
print(test_word.count("abcabc"))

1


With `count()`, getting Santa’s destination floor becomes pretty short:

In [85]:
floor = raw_input.count("(") - raw_input.count(")")
print(f"Santa arrived on this floor: {floor}")

Santa arrived on this floor: 280


### Part two

#### The challenge

> Now, given the same instructions, find the position of the first character that causes him to enter the basement (floor -1). The first character in the instructions has position 1, the second character has position 2, and so on.
> 
> For example:
> 
> - `)` causes him to enter the basement at character position 1.
> - `()())` causes him to enter the basement at character position 5.
>
> What is the position of the character that causes Santa to first enter the basement?

#### The strategy

We used the `for` loop code from part one as the basis for our solution for part two, but with a couple of changes:

* As we go through each character in the input, we need to keep track of what character we’re on. As soon as you hear “keep track of”, you should be thinking “we need a variable for that!”
* We can determine the current floor by calculating `open_parentheses - close_parentheses`. If this value is `-1`, we’re at the basement level! We can print the current position and exit the loop with the `break` statement.

#### The solution

In [86]:
open_parentheses = 0
close_parentheses = 0
current_position = 1

for character in raw_input:
    if character == "(":
        open_parentheses += 1
    elif character == ")":
        close_parentheses += 1
    else:
        print("Something's wrong with input.")
        print(f"Unexpected character: {character}")
        quit()
        
    # See if we’re at the basement level (level -1)
    current_floor = open_parentheses - close_parentheses
    if current_floor == -1:
        print(f"Santa found the basement at position: {current_position}")
        break
        
    current_position += 1

Santa found the basement at position: 1797


## The “Day One” challenge from the 2020 Advent of Code

We also tackled both parts of the Day One challenge of the 2020 Advent of Code.

### Part one

#### The challenge

Here’s the challenge, taken straight from [its webpage](https://adventofcode.com/2020/day/1):

> Before you leave, the Elves in accounting just need you to fix your expense report (your puzzle input); apparently, something isn't quite adding up.
>
> Specifically, they need you to find the two entries that sum to 2020 and then multiply those two numbers together.
> 
> For example, suppose your expense report contained the following:
> 
> 1721
> 979
> 366
> 299
> 675
> 1456
>
> In this list, the two entries that sum to 2020 are 1721 and 299. Multiplying them together produces 1721 * 299 = 514579, so the correct answer is 514579.
>
> Of course, your expense report is much larger. Find the two entries that sum to 2020; what do you get if you multiply them together?

In [87]:
1140
1736
1711
1803
1825
1268
1651
2007
1923
1661
1788
1876
2003
1752
1988
1955
1568
1478
1699
1717
1828
1636
1387
1870
1658
1572
1703
1185
1569
1515
1142
1407
1587
1608
1827
1546
1808
1937
1815
1957
1401
1763
1970
1960
1853
1987
1865
1567
1664
1961
1771
1846
1971
1416
1897
633
1708
1606
515
1397
1873
1374
1969
1918
1170
1660
1494
1764
2002
1938
1396
1926
1714
1659
1805
1593
1899
1850
1644
1877
1561
1895
1985
1353
395
1919
1522
1745
1721
901
1765
1939
2009
1949
1852
1792
1749
1675
1883
1240
1868
1615
1693
1720
1388
1325
1337
867
1751
1408
1715
1942
1706
1894
1260
1945
1700
1148
1373
351
1790
1861
1755
1155
1622
1743
1872
1979
1262
1789
1305
1311
1729
1929
823
1623
2005
1932
1814
1909
1728
1592
1712
1363
1338
1804
1402
1198
264
1117
1791
1419
1229
1924
1838
1785
1982
1683
1950
1199
1984
1830
1921
1980
1834
1341
1282
1989
1854
1395
1847
1900
1913
1777
1779
1333
1800
1966
1543
1882
1375
1811
1673
1679
889
1670
1879
1312
1741
1772
1663
1776
1642
1674
1472
1580
1264
1738
1999
1637

1637

#### Importing the input into Python

The simplest way to bring the input into Python is to copy the input, paste it into a triple-quoted string, and assign the string to a variable, which I named `raw_input`:

In [88]:
raw_input = """1140
1736
1711
1803
1825
1268
1651
2007
1923
1661
1788
1876
2003
1752
1988
1955
1568
1478
1699
1717
1828
1636
1387
1870
1658
1572
1703
1185
1569
1515
1142
1407
1587
1608
1827
1546
1808
1937
1815
1957
1401
1763
1970
1960
1853
1987
1865
1567
1664
1961
1771
1846
1971
1416
1897
633
1708
1606
515
1397
1873
1374
1969
1918
1170
1660
1494
1764
2002
1938
1396
1926
1714
1659
1805
1593
1899
1850
1644
1877
1561
1895
1985
1353
395
1919
1522
1745
1721
901
1765
1939
2009
1949
1852
1792
1749
1675
1883
1240
1868
1615
1693
1720
1388
1325
1337
867
1751
1408
1715
1942
1706
1894
1260
1945
1700
1148
1373
351
1790
1861
1755
1155
1622
1743
1872
1979
1262
1789
1305
1311
1729
1929
823
1623
2005
1932
1814
1909
1728
1592
1712
1363
1338
1804
1402
1198
264
1117
1791
1419
1229
1924
1838
1785
1982
1683
1950
1199
1984
1830
1921
1980
1834
1341
1282
1989
1854
1395
1847
1900
1913
1777
1779
1333
1800
1966
1543
1882
1375
1811
1673
1679
889
1670
1879
1312
1741
1772
1663
1776
1642
1674
1472
1580
1264
1738
1999
1637"""

Let’s take a look at `raw_input`:

In [89]:
raw_input

'1140\n1736\n1711\n1803\n1825\n1268\n1651\n2007\n1923\n1661\n1788\n1876\n2003\n1752\n1988\n1955\n1568\n1478\n1699\n1717\n1828\n1636\n1387\n1870\n1658\n1572\n1703\n1185\n1569\n1515\n1142\n1407\n1587\n1608\n1827\n1546\n1808\n1937\n1815\n1957\n1401\n1763\n1970\n1960\n1853\n1987\n1865\n1567\n1664\n1961\n1771\n1846\n1971\n1416\n1897\n633\n1708\n1606\n515\n1397\n1873\n1374\n1969\n1918\n1170\n1660\n1494\n1764\n2002\n1938\n1396\n1926\n1714\n1659\n1805\n1593\n1899\n1850\n1644\n1877\n1561\n1895\n1985\n1353\n395\n1919\n1522\n1745\n1721\n901\n1765\n1939\n2009\n1949\n1852\n1792\n1749\n1675\n1883\n1240\n1868\n1615\n1693\n1720\n1388\n1325\n1337\n867\n1751\n1408\n1715\n1942\n1706\n1894\n1260\n1945\n1700\n1148\n1373\n351\n1790\n1861\n1755\n1155\n1622\n1743\n1872\n1979\n1262\n1789\n1305\n1311\n1729\n1929\n823\n1623\n2005\n1932\n1814\n1909\n1728\n1592\n1712\n1363\n1338\n1804\n1402\n1198\n264\n1117\n1791\n1419\n1229\n1924\n1838\n1785\n1982\n1683\n1950\n1199\n1984\n1830\n1921\n1980\n1834\n1341\n1282\n1989\

Remember, characters preceded by a backslash (`\`) are . [*escpaed characters*.](https://www.w3schools.com/python/gloss_python_escape_characters.asp) `\n` characters are [*newline characters*.](https://www.computerhope.com/jargon/n/newline.htm)

We’ll use the newline characters as the [delimiter](https://www.computerhope.com/jargon/d/delimite.htm) to split `raw_input` into an array:

In [90]:
split_input = raw_input.split("\n")

Here’s the result:

In [91]:
split_input

['1140',
 '1736',
 '1711',
 '1803',
 '1825',
 '1268',
 '1651',
 '2007',
 '1923',
 '1661',
 '1788',
 '1876',
 '2003',
 '1752',
 '1988',
 '1955',
 '1568',
 '1478',
 '1699',
 '1717',
 '1828',
 '1636',
 '1387',
 '1870',
 '1658',
 '1572',
 '1703',
 '1185',
 '1569',
 '1515',
 '1142',
 '1407',
 '1587',
 '1608',
 '1827',
 '1546',
 '1808',
 '1937',
 '1815',
 '1957',
 '1401',
 '1763',
 '1970',
 '1960',
 '1853',
 '1987',
 '1865',
 '1567',
 '1664',
 '1961',
 '1771',
 '1846',
 '1971',
 '1416',
 '1897',
 '633',
 '1708',
 '1606',
 '515',
 '1397',
 '1873',
 '1374',
 '1969',
 '1918',
 '1170',
 '1660',
 '1494',
 '1764',
 '2002',
 '1938',
 '1396',
 '1926',
 '1714',
 '1659',
 '1805',
 '1593',
 '1899',
 '1850',
 '1644',
 '1877',
 '1561',
 '1895',
 '1985',
 '1353',
 '395',
 '1919',
 '1522',
 '1745',
 '1721',
 '901',
 '1765',
 '1939',
 '2009',
 '1949',
 '1852',
 '1792',
 '1749',
 '1675',
 '1883',
 '1240',
 '1868',
 '1615',
 '1693',
 '1720',
 '1388',
 '1325',
 '1337',
 '867',
 '1751',
 '1408',
 '1715',
 '1942

#### Building lists with list comprehensions

[**List comprehensions**](https://jakevdp.github.io/WhirlwindTourOfPython/11-list-comprehensions.html) are a way of building new lists from existing ones, using a syntax that’s similar to `for` loops.

Here’s an example that uses a list comprehension to build a list whose values are double those of the original:

In [92]:
my_list = [1, 2, 3, 4, 5]
double_list = [x * 2 for x in my_list]
print(double_list)

[2, 4, 6, 8, 10]


Here’s another example, where we use a list comprehension to create a list contaning uppercase versions of the items in the original:

In [93]:
my_strings = ['apple', 'banana', 'cantaloupe']
my_all_caps_strings = [x.upper() for x in my_strings]
print(my_all_caps_strings)

['APPLE', 'BANANA', 'CANTALOUPE']


We used a list comprehension to make a new list, which we’ll call `expenses`, based on `split_input`. The resulting list contained the contents of `split_input`, converted into integers:

In [94]:
expenses = [int(string) for string in split_input]

Let’s get a look at `expenses`:

In [95]:
expenses

[1140,
 1736,
 1711,
 1803,
 1825,
 1268,
 1651,
 2007,
 1923,
 1661,
 1788,
 1876,
 2003,
 1752,
 1988,
 1955,
 1568,
 1478,
 1699,
 1717,
 1828,
 1636,
 1387,
 1870,
 1658,
 1572,
 1703,
 1185,
 1569,
 1515,
 1142,
 1407,
 1587,
 1608,
 1827,
 1546,
 1808,
 1937,
 1815,
 1957,
 1401,
 1763,
 1970,
 1960,
 1853,
 1987,
 1865,
 1567,
 1664,
 1961,
 1771,
 1846,
 1971,
 1416,
 1897,
 633,
 1708,
 1606,
 515,
 1397,
 1873,
 1374,
 1969,
 1918,
 1170,
 1660,
 1494,
 1764,
 2002,
 1938,
 1396,
 1926,
 1714,
 1659,
 1805,
 1593,
 1899,
 1850,
 1644,
 1877,
 1561,
 1895,
 1985,
 1353,
 395,
 1919,
 1522,
 1745,
 1721,
 901,
 1765,
 1939,
 2009,
 1949,
 1852,
 1792,
 1749,
 1675,
 1883,
 1240,
 1868,
 1615,
 1693,
 1720,
 1388,
 1325,
 1337,
 867,
 1751,
 1408,
 1715,
 1942,
 1706,
 1894,
 1260,
 1945,
 1700,
 1148,
 1373,
 351,
 1790,
 1861,
 1755,
 1155,
 1622,
 1743,
 1872,
 1979,
 1262,
 1789,
 1305,
 1311,
 1729,
 1929,
 823,
 1623,
 2005,
 1932,
 1814,
 1909,
 1728,
 1592,
 1712,
 1363,

How many items are in this list?

In [96]:
print(len(expenses))

200


If there are 200 items in the list, that means that there are 200 * 199 possible combinations that could sum up to 2020. That’s almost 40,000 combinations!

#### Using the `pop()` method to retrieve and remove the last item of a list

Here’s an example:

In [97]:
my_numbers = [1, 2, 3, 4, 5]
last_number = my_numbers.pop()
print(f"We popped this number from the end of the list: {last_number}")
print(f"Here’s what the list looks like now: {my_numbers}")

We popped this number from the end of the list: 5
Here’s what the list looks like now: [1, 2, 3, 4]


#### The strategy

The approach that our initial solution took went like this:

* Pop the last number from the list, and hold onto it.
* Try to add it to every other number in the list, checking to see the two add up to 2020. If you find another such number, stop.
* If, after trying to add that popped number to every other number in the list, you don’t create a pair that sums to 2020, it’s clear that the popped number isn’t one of the two numbers in the list that adds up to 2020. Go back to the first step and try again.

In this process, we reduce the size of the list by one.

Here’s our first solution:

In [98]:
last_number = expenses.pop()
solved = False

while not solved:
    
    for expense in expenses:
        if expense + last_number == 2020:
            print(f"The 2 numbers are {expense} and {last_number}.")
            solved = True
            break

    last_number = expenses.pop()


The 2 numbers are 1387 and 633.


### Part two

#### The challenge

> The Elves in accounting are thankful for your help; one of them even offers you a starfish coin they had left over from a past vacation. They offer you a second one if you can find three numbers in your expense report that meet the same criteria.
> 
> Using the above example again, the three entries that sum to 2020 are 979, 366, and 675. Multiplying them together produces the answer, 241861950.
> 
> In your expense report, what is the product of the three entries that sum to 2020?

There *is* a way to do this from first principles, but it seems like a lot of work. Isn’t there a better way to do this?

Luckily, this is Python, which comes with a whole lot of libraries, including the [itertools](https://realpython.com/python-itertools/) library, which contains functions for generating *combinations*.

To make the functions in itertools available to your programs, you use Python’s `import` statement. The following imports *all* of itertools’ functions into your programs:

In [99]:
from itertools import *

One of the functions that itertools provides is the `combinations()` function, which generates combinations from items in a list.

Here’s a demonstration of `combinations()` that generates all the possible 2-item combinations of numbers in a list.

In [100]:
mu_numbers = [1, 2, 3, 4, 5, 6]
my_combos = list(combinations(my_numbers, 2))
print(my_combos)

[(1, 2), (1, 3), (1, 4), (2, 3), (2, 4), (3, 4)]


The combinations are [**tuples**](https://realpython.com/python-lists-tuples/#python-tuples), which are ordered collections of objects. Unlike lists, tuples are *immutable* — you can’t alter their items, or add or remove items from them.

We used `combinations()` to create a list of all the possible combinations of three numbers in the expenses list:

In [101]:
expense_3_item_combos = list(combinations(expenses, 3))
print(expense_3_item_combos)

[(1140, 1736, 1711), (1140, 1736, 1803), (1140, 1736, 1825), (1140, 1736, 1268), (1140, 1736, 1651), (1140, 1736, 2007), (1140, 1736, 1923), (1140, 1736, 1661), (1140, 1736, 1788), (1140, 1736, 1876), (1140, 1736, 2003), (1140, 1736, 1752), (1140, 1736, 1988), (1140, 1736, 1955), (1140, 1736, 1568), (1140, 1736, 1478), (1140, 1736, 1699), (1140, 1736, 1717), (1140, 1736, 1828), (1140, 1736, 1636), (1140, 1736, 1387), (1140, 1736, 1870), (1140, 1736, 1658), (1140, 1736, 1572), (1140, 1736, 1703), (1140, 1736, 1185), (1140, 1736, 1569), (1140, 1736, 1515), (1140, 1736, 1142), (1140, 1736, 1407), (1140, 1736, 1587), (1140, 1736, 1608), (1140, 1736, 1827), (1140, 1736, 1546), (1140, 1736, 1808), (1140, 1736, 1937), (1140, 1736, 1815), (1140, 1736, 1957), (1140, 1736, 1401), (1140, 1736, 1763), (1140, 1736, 1970), (1140, 1736, 1960), (1140, 1736, 1853), (1140, 1736, 1987), (1140, 1736, 1865), (1140, 1736, 1567), (1140, 1736, 1664), (1140, 1736, 1961), (1140, 1736, 1771), (1140, 1736, 1846),

Now that we have these combos, how do we find their sum?

Luckily, Python has a built-in `sum()` function that returns the sum of the numbers in a given tuple or list:

In [102]:
print(sum((2, 3)))
print(sum([7, 8, 9, 10]))

5
34


We can use `sum()` in a `for` loop to find the combo whose numbers add up to 2020:

In [119]:
for combo in expense_3_item_combos:
    if sum(combo) == 2020:
        print(f"These are the 3 numbers that add to 2020: {combo}")
        break

In [118]:
867 + 264 + 889

2020

In [105]:
numbers = [-5, 5, 8, -1.2, 90, 33, -12]

In [106]:
postive_numbers = [number for number in numbers if number >= 0]
print(postive_numbers)

[5, 8, 90, 33]


In [107]:
double_positive_numbers = [number * 2 for number in numbers if number >= 0]

In [108]:
double_positive_numbers

[10, 16, 180, 66]

In [109]:
resto1 = {
    "name": "Rooster & the Till", 
    "rating": 4.5, 
    "price": 3, 
    "delivery": True, 
    "takeout": True
}

In [110]:
resto1

{'name': 'Rooster & the Till',
 'rating': 4.5,
 'price': 3,
 'delivery': True,
 'takeout': True}

In [111]:
resto1['name']

'Rooster & the Till'

In [112]:
resto1['name'] = "Rooster and the Till"

In [113]:
resto1

{'name': 'Rooster and the Till',
 'rating': 4.5,
 'price': 3,
 'delivery': True,
 'takeout': True}

In [114]:
resto1['gross_bathrooms'] = False

In [115]:
resto1

{'name': 'Rooster and the Till',
 'rating': 4.5,
 'price': 3,
 'delivery': True,
 'takeout': True,
 'gross_bathrooms': False}

In [116]:
resto1['hipster_joint']

KeyError: 'hipster_joint'

In [ ]:
resto1.get('takeout')

In [ ]:
resto1.get('hipster_joint')

In [ ]:
restaurants = [
    {"name": "Rooster & the Till", "rating": 4.5},
    {"name": "Ella's", "rating": 4},
    {"name": "Trip's Diner", "rating": 3.5}
]

In [ ]:
restaurants

In [ ]:
restaurants[1]

In [ ]:
restaurants[1]['rating']

In [ ]:
restaurants[2]['name']

In [ ]:
resto1

In [ ]:
del resto1["gross_bathrooms"]

In [ ]:
resto1

In [ ]:
resto1.keys()

In [ ]:
resto1.values()

In [ ]:
list(resto1.keys())

In [ ]:
list(resto1.values())

In [ ]:
for key in resto1.keys():
    print(f"The key '{key}' has the value {resto1[key]}.")

In [ ]:
resto1['name']

In [ ]:
resto1

In [ ]:
resto1['categories'] = ["American", "small plates", "bistro"]

In [ ]:
resto1

In [ ]:
resto1["categories"][1]

In [ ]:
for category in resto1["categories"]:
    print(category)

In [ ]:
is_hipster = "bistro" in resto1["categories"]

In [ ]:
is_hipster

In [ ]:
restaurants

In [ ]:
really_good_restaurants = [restaurant for restaurant in restaurants if restaurant["rating"] >= 4]

In [ ]:
really_good_restaurants

In [ ]:
restaurants

In [ ]:
first_half_of_phone_book = [restaurant for restaurant in restaurants if restaurant["name"] < "M"]

In [ ]:
first_half_of_phone_book

In [ ]:
"Alice" < "Bob"

In [ ]:
"Alice" > "Bob"

In [ ]:
"A" > "a"

In [ ]:
asc("A")

In [ ]:
chr("a")

In [ ]:
ord("A")

In [ ]:
ord("a")

In [ ]:
"alice" > "BOB"

In [ ]:
raw_input = """3-5 f: fgfff
6-20 n: qlzsnnnndwnlhwnxhvjn
6-7 j: jjjjjwrj
8-10 g: gggggggggg
5-6 t: ttttttft
6-11 h: khmchszhmzm
4-6 q: qqbjqqqj
6-8 j: gtkwwjjj
13-14 w: lwgwrwczwwhkww
2-4 p: vcnptzdppvpdzp
4-6 n: pmlnzwrnbnxnpmwmhbnz
4-9 w: wwprdwwwx
13-20 b: rbbcbbbbbbbbsbbbbbmj
1-3 r: zwhrnvr
4-6 f: fmfgfcf
13-15 z: zxzzzzzzzzvxkplvzzz
12-18 r: phjqccgmgzmdwxgtrtb
5-19 s: ssssssssssssssssssss
6-7 d: pddqmdd
7-9 p: pwrlpvlhpfpppzp
4-5 m: mpmmmm
9-10 s: ssxsssssrj
6-9 p: lpjrvpppp
6-13 r: rrrzvtrgrhqxqrvrvwzr
1-3 p: dpjp
6-7 w: wkjnmww
1-5 r: cfqhhr
5-9 m: wmmlpgmqmmrm
12-13 x: sxgnwxxkjwrxr
6-8 n: lnnrnnnnn
10-11 q: kqqqzqqfqqqqq
12-17 k: kkkkkkkkkkpkkhgvnjhk
12-13 v: vvvvvvvvvvvvvv
12-15 h: hhhhhhhjkhhhhhth
6-7 l: hmqtlllnllnlmtqnn
6-9 m: mcmmmdnmh
3-5 k: kkhtn
2-4 n: tnfjmnnn
5-6 j: cjwlmn
4-11 b: bcbbkrlgcbbc
5-10 r: rcrcgrrrzwrrxgpzwrcm
3-4 n: znnn
11-14 s: sssjsssssssssss
6-7 d: rwdddddmdddddkdd
1-4 s: gssssskssss
4-5 j: jjjjzj
11-12 j: cjjjgpjxnjjk
3-6 z: zzkdzzz
7-9 d: mkddddqdhddgdd
1-9 d: rxwqcdrswdg
2-4 m: gmmmn
2-4 k: kkkmk
15-16 r: rrrrrrrrlrrrrrrr
1-11 g: grgggcgggcggp
7-8 q: qqqjqqgq
13-14 c: cxccccgccvcpccccc
2-5 j: txvwlls
3-13 w: plwqwhbwdgxcwfmwjl
9-11 g: ggggggggggggggggdgg
3-12 g: ggsggggggpgm
3-11 g: bcgqgxmbjpwm
8-9 z: zxvzrzzzzzdvzzgz
17-18 z: zzzxzzzzzzzzzzzzsw
9-10 t: ftrmtttktttttdtmdk
15-17 r: rjrrprrrrrrrrrrfrrrr
7-8 k: kkkkkkkkk
6-13 n: nnnvnfgntnnnjnnxh
17-19 l: llllflllllllllllqlrl
11-13 d: hdlddddddhxddddkd
13-17 h: hhhhhhhhhhhhnhhhjhhh
1-12 c: ncccczwcnctcwcc
4-5 w: wgwkl
12-13 w: wwwwwwwwwwwwfwww
3-12 s: scssgwshsspsss
15-17 w: wwwwwcwxmhwwwwwwwww
1-3 x: xxxxw
8-9 p: mpppppppzptp
11-12 n: nntnbnnnnnmx
2-10 x: xrspbmkcthqsdxrdxwx
13-14 l: lllllllpllllll
5-10 x: xxxxjxjdrx
7-8 k: kkkkkkjk
9-11 k: vkkkkkkvkkzddbzr
3-8 r: lbrrcwbdf
6-14 l: lgwllrgllllllgllll
6-9 q: nqqkhqqtqgqc
2-4 g: ggxggggggg
12-13 k: kkkkkjkkkkkvf
5-9 b: scvbbpzbbbzbfb
16-17 d: mdddddddddddddcttd
4-10 l: lllflllllllllllllll
3-6 q: qqqqqtqqqq
1-3 n: pblsghl
9-11 v: bvvvmvfvvgtvfpv
4-5 s: ssshq
5-8 g: ggwgnggp
1-4 f: ffmr
4-5 g: ggwzs
4-5 h: fhhqs
10-17 l: clhlllhslxpgljpvlrkl
7-9 m: pnswhtmvmsrmjwrbfz
6-16 v: vvvvqcvvvdzvjvldvv
5-6 c: cntcrl
12-13 n: nnnnnnnnnnnnln
2-4 w: wwww
2-13 w: swrqssmmwrxtw
15-16 z: zzzzzzzzzzzzzzgqzzz
6-8 c: cctvkbdcwcbvhc
1-2 w: wwzdwjtm
1-3 z: mzzh
6-13 t: fttttdwtttttkht
3-9 g: cggfjgqngwmj
1-5 k: pkkkkkkrk
1-5 q: qknqd
7-11 t: tttttckttkb
6-7 q: qqqqqbl
16-17 c: ccccccccccccccctc
2-5 v: vfcnvlvvvxvrnvvvvvv
1-5 w: nwtwnnwwhtwwdwwww
2-10 v: ftvxxbjzlqctp
5-7 d: nddddgddjdbk
8-11 s: wxqjwkcsflssm
17-20 v: vvvvvvvvpvgvvvvvvvvv
6-8 w: wwlwwwqfw
13-14 r: rrrrgrrrrrrrrr
5-6 t: ttttzq
13-14 c: cccccccccccccc
11-12 k: kwzkkkbdkjkk
10-11 t: ttttpttttrwwtttt
3-4 w: ljwwww
7-8 q: ggqzdvqqw
1-3 z: zzsg
1-2 c: crjvdhgwckszmzpcjmr
10-11 n: nngnntnnnqn
7-16 w: bwdgwwwgwwnnqtcwc
4-16 v: vfkvqvvsvnjhfvvv
10-11 p: fgbhpzqvkmn
10-11 w: wswgwqwszwwhwnwww
15-16 t: stwtfxttthptttttt
1-3 v: vlvsdmrv
5-8 z: zzzzfzzz
14-15 t: ttpvtrlqtlcdrlv
2-6 b: qwfnbkc
6-9 p: jppjpspplm
3-4 s: brgqssz
2-6 q: cqwqlq
13-14 d: rrdljqdddddcxzdc
8-13 s: tsnssssssssrcssrssq
7-8 d: dwddddgl
9-15 b: gspmhlgbbzbbbrbt
3-4 h: khhh
3-4 c: nkccpmcct
6-7 g: ggggggg
17-18 p: ppppppppppppppppjnp
2-6 f: rfjxtff
3-4 g: ggvg
7-9 k: kkkkkkkkkkk
11-16 f: fjdffrtfffjfdffcln
4-8 w: lhpdwbnfssswwfswwwrw
5-9 l: lvlllflll
5-6 t: wtrtlr
4-8 m: mnmmjgqtgs
5-11 n: nnrngbnntnk
5-8 s: ssssfssss
7-14 m: hrnbhsfcvdmxbmvzfvnq
1-5 x: xxxxxxxxwxxdg
1-10 v: vcjvvjvvvv
5-10 l: vvdmhskmprszklvvl
3-17 d: dsddkddddddqddddfjdd
9-13 p: pppppppptpppqp
12-13 h: hhhhhhhhhhjhh
1-4 z: mrvlzlzhzlp
3-4 h: hhmc
7-17 w: wwwwwwwwwwwwwwwwww
6-8 z: cgzxzznzzdhcvwh
4-5 g: mgggg
7-8 t: ttrthtdkjtgtts
8-9 w: wwwwwwwpw
9-15 k: kkkkkkkkkkkkkkk
8-14 q: qqqqqqqqnqqqqq
14-19 w: jkwbwwlwdcmwmmwwrwr
4-5 m: hsxgkdmz
2-4 v: vsvqvcv
4-9 h: grhnhhmhhxhlh
6-14 t: tmttrjbtjtfttc
1-4 d: pjdbrb
15-16 r: rrrrrblrrrrrrrxwr
16-18 p: pppjpppppptpplpppppp
4-7 x: mxtngxnpsx
9-10 b: bbbdbqvmrgbzbbbb
4-5 w: wwwwj
1-2 d: dgws
3-10 l: lglsphlbgfmgdtw
10-11 h: hhhhhfhhhpdhh
5-9 x: xxxxqgxxjf
11-19 m: bfmxmmfgzzhqwsmgwmlr
10-14 x: xxxrxkxxqrlxmpxbdx
3-8 l: lxlxvwlllznlptfv
5-9 h: hhthhhhshdq
1-5 q: sqqqqqqq
9-14 j: jxfcjpjfjjgjjjjjjjhj
1-7 g: gvdgjgsvvgnd
2-5 t: tmttpttttt
17-19 c: ccccclcccccccccccccc
10-13 d: mdvddtsddrxtx
11-12 g: gggggggwggzgp
3-5 k: hzxnpkq
9-11 q: qqqqqsqqqqq
8-9 x: btrkbxngx
3-5 g: jpkchcxcxhzfhsggqkg
7-13 s: ssssssssscssssn
13-15 p: pppppppnppppppp
5-15 f: ffffjfffffffffgf
2-4 p: zqphp
1-4 r: wrrr
3-9 b: jsjcbrkkczkzmjbg
1-2 k: kkkkl
2-3 n: dlnxjwzstsdxns
7-14 v: fvvvvzhhvpsvvckdvv
2-8 r: rrrrwrhrrrr
5-8 n: nnxdnnnnnn
5-12 c: scmcdscccccmcc
9-10 v: vvvvvvvvjm
14-15 w: lwppdwwwwdmrrww
6-7 c: qcgcvxdrcccpxchrnlq
8-10 x: hbrqsksxwxxttz
18-19 w: jnftgwpwmwfdgrcpkww
3-4 k: ffkj
5-9 f: flffsvffsff
3-4 k: kkkkk
1-3 l: dlglkll
10-16 f: bfffjzfffnfffffxhfff
11-20 d: ddbdqsddddddddddfddd
6-20 l: vmjltmkclbmqrflzgdzl
5-19 f: ffffffffffffffffffff
5-10 n: nnnnmnnfnz
4-13 f: kmrfrfkvjxdbftvhnsdm
10-11 k: kkkkkkkkskk
12-13 k: kkkkkkkkkkkhj
7-8 p: ppppppnhw
5-8 l: gtpllwklh
13-14 h: hhhhhhhhzhhwsrh
7-15 h: hhhhhhthhhhhthqgfh
1-3 f: mfdmhv
15-16 f: fffffflfffffffbqf
12-13 c: cccmcccccccxn
2-12 q: nqltxqfdcrxqvt
8-10 d: dddddddddvdddd
2-9 g: mdrwnvtsd
12-13 l: lllllllllllll
2-3 m: mnrgmdm
6-9 v: hvvvvvvvv
4-7 q: bqqqrqcqqqq
8-11 v: vvvvdvvvvvr
3-4 n: cnnw
2-12 r: rrrklfrrrksrr
2-12 p: pxppklhsppwdxwcpzvm
10-11 n: nnnnnnnnnnn
5-6 w: wfwwww
3-5 p: pppppp
3-8 g: ggggggggggggggg
1-4 x: xxxxx
11-16 d: ngcdkglddtppbddgdrd
11-12 k: kkkkkkkkkkhr
12-13 c: hgxxchcvxpdlsrt
1-3 m: vmmm
4-6 z: kvzgzzzp
11-12 d: dzddzlcbvdxk
10-11 g: ggsgggggggggdz
6-8 q: xfhgpqltbfbdzqg
4-7 v: vvvlvcxmvvxq
5-7 v: vkkgvgvnpvvlpgkv
6-7 b: bbkvnwbqblbvbb
2-4 l: wmll
6-9 j: djqjcxxljm
3-4 j: nmzcsnlnjjjdms
3-8 w: wnwwsvww
9-20 m: wbntxzztwmblxmsmltmg
7-8 r: rprrrtrf
1-3 r: lnrfxfswmhgvf
5-12 k: zlkdlktsrqjt
13-14 v: vvvvvvvvvvvvvbv
11-12 d: ddddddddddgc
9-11 c: mdwpjjcdcrc
5-6 m: mmmmnl
5-6 l: lblldn
13-14 c: sccfcxsnsclccc
4-5 q: qqkkq
5-8 m: mmmmmmmm
5-8 z: tzhzzttzvwfcv
12-15 b: bbbbbbbbbbbtdfb
7-10 b: bbnplbbxxbh
1-7 x: zxxxxxvx
6-17 b: kbnbbbbbptbbrbbbb
1-3 l: llllll
13-15 c: ccccccccccccqcz
1-8 s: sssssssb
12-19 x: xxxxxxxxpxxcxxxxxxq
12-13 x: xxxxxxxxxxxxxx
14-18 w: wwwwwwwwwwwwwlwmbk
7-8 q: mzwqblqqz
1-7 p: hrppppgdzp
2-7 n: qmvpmsjncgkgpbb
8-10 x: xxxxxxxwxx
2-4 v: rfcpmpvsswsrjkxpdrxh
14-16 k: kkkkkkkkkkkhkkskkt
14-15 d: ddddddkdddddrqdd
4-16 m: njmqmmmjfmmxhmwgdbc
12-13 h: hdmdfhhhzhkdhr
3-14 n: nbntrsmnmrcwbf
6-8 h: hhwhhlhnhh
18-19 v: vvvvvvvvvvvvjvvvvvp
11-20 l: ldllllbvllmqllmlllfl
2-3 d: dcddd
4-6 n: bsnncn
3-6 f: dwlfqfzgs
8-12 d: xdddddfhnddfddd
5-13 l: llllqlllllmllll
10-14 m: mnmzrmmmmwmmqm
3-12 c: qjchcclnbccccpc
7-16 j: jjjjjjwjjjjjjrjnjj
1-10 v: svvdvvgmgpkfkvhvv
2-4 b: gbpbczblbbv
7-9 w: drwwbwtgwswwww
2-8 h: fhkprhplz
1-6 m: mmmmwm
3-4 g: gggp
1-3 q: qqqqj
10-11 h: htfrdwrnkzntnfpdmkhq
5-6 x: nxxxxxxxj
7-8 w: wwwcwwww
1-5 h: lhhhhh
7-8 f: ffjfftff
3-5 g: kdgfgjgbghggp
2-11 p: flgpvfbhpzxmmntw
12-13 z: zzzzzzhzzzzgzzqz
12-13 j: ttmhmdfjpdpnj
5-7 h: hhhhvhh
9-13 x: xhxbxxsxxqwxxvxxxx
1-3 c: ccchcc
2-5 h: hhrsh
1-3 q: rqqq
2-8 x: sxrxbcxx
12-15 b: bbbbnbbrbcbtbbb
14-18 m: hmmmmmmmkmmmmmmmmrmm
1-4 v: vvvv
5-11 h: brnzvqhrrwh
4-10 f: dbtfdvlfnffqmwp
3-16 r: krrrvmrwrrrtrrrrkrwr
11-15 c: gxbccccccccccbh
13-18 z: zzrzzzzzzzzzzzzzzzz
1-3 l: lkvxtrthfvmdzn
7-9 p: vpppmpppppp
3-4 w: wwwwwwww
8-11 z: vlzzzlzvzzmzzz
2-4 v: vqfv
10-11 s: sssssslssgss
4-5 z: zzzww
11-19 p: zpbpdjfjpdfgnpphhpg
1-8 f: fftlxmfl
9-10 p: phlpqzppxpxp
5-7 q: kbqjqqq
3-6 m: jcblmv
9-10 l: llllllllll
13-14 w: wbwhwwwwhwqwmswwmjwf
6-11 j: jjjjvtjwjbdjj
4-8 j: jljjjvgj
3-4 r: lrhvrzrr
2-4 p: plpfkphpq
2-6 b: bbbbrbvwbbgh
2-11 t: mmjmcwcdcbtztxfbtst
7-15 r: grsrrjrrtrgvrtrrr
6-17 d: dddddgddpddzddddddd
3-10 t: qpnbrnmmjtjxtdg
6-7 v: vvvvvhb
5-20 j: vxmkjnssjdtldsdwltlg
7-10 r: rrrrrrrrrrr
2-4 r: rrpl
3-4 m: mmzg
1-8 h: fshhjrhm
5-8 m: mmmvmmvmmnmdxnjcjpn
8-10 m: mmmmmmmlmzmm
14-15 q: qqqqqqqqqqqqqqb
6-16 q: qhkdtqnzqqqjgjrqrkbq
3-9 b: gtbbfgfhbs
14-16 h: hhhhdhhhmhhhhmhhh
9-10 n: nnnndnnnvjnl
11-12 p: pqpppppppppppp
7-9 d: ddpwfdtdd
6-10 z: qzzzzjzzmz
9-14 q: qqdjqdqqrqqwqnqqp
6-7 k: kkkkkbkk
7-14 j: hjvjmzkhgltffljjvjj
4-5 z: znhzz
11-16 r: rrrrbxdrrqrrsrrb
5-8 v: vvhvvvvvwvlvhlqg
4-10 n: wvnnjnndcnk
5-6 g: ggkggk
8-9 w: wwwwwwwww
5-9 r: mrwrrrrrx
4-6 r: jcfrjrrrg
6-9 w: wkwwsxwgfcwqb
5-6 x: grxjxnrb
2-7 s: nsmldnsszdk
5-9 j: gjjzkjvjwjcjbj
6-11 r: jtrgcvwccrrrfmj
7-12 j: jfjjmjjjjxjqzdjjxq
1-2 n: nnwqcnr
5-16 l: lllpclllllllblllll
5-6 j: wjgdjcdpjxfjxnjj
3-16 s: klstwqksfclmcbbcn
17-18 t: ttttdtttttthtttttttt
9-13 p: ppppppvpvhhqp
1-2 n: tsnnn
3-4 p: jpqpjmpplpw
4-6 t: dtfthsxtfpd
2-6 m: zhszctkm
1-13 p: dppmkpbprgppxcpvhrpz
2-7 m: gvmmflp
9-16 c: rcdsppjbcwdxnfbc
5-15 p: pppxpppqnhpptpppz
10-13 q: xfhqhkqkjqjwqqcfrgl
14-15 t: tttttttttttttdvtttt
5-8 q: qqqbqvqzqqn
1-7 v: cvgvvbcgjv
6-9 r: rrrrrrrdr
2-11 j: wjjgxngswkj
11-13 d: dddddddddddddm
5-8 p: pzppppppkjp
15-16 c: cccccccccccccctc
1-7 c: vsvkccbc
2-5 h: hhhhchhhhhhhhhh
9-11 n: nnnnnnnnnnnnn
11-12 q: mdmfmkxcxzjq
6-11 m: mmkgnjdhfbc
1-4 d: dgrdwbdfdkdrmthsj
12-14 k: kkkkkkkkkrjkkkkk
2-4 d: hdrvdzd
2-3 s: sxmsssssssssssssssss
13-15 z: zzzzzzlzzlzztzzz
3-8 r: lxrrrnvrtrgnmkrr
6-7 w: wwlbrwwww
4-7 r: zlnmmkpnrkkcrrxrmfq
2-6 v: zgvvjjvvjhjv
8-11 l: llllllllllcl
8-14 m: mlmldmmgnqzmmmm
7-10 z: wzzzzhtzdqzznzz
4-7 k: mvgwkkk
5-9 p: pdpltvdpptpppg
4-15 q: qqqfqbzqlqqqqqpqfqs
5-7 q: qqqcxqqm
14-17 x: rltjxxbxxxllsxwcx
1-13 w: gwggwwwwwwbsbwkwwd
8-14 k: kkbkkkkhkkkxjzblgk
5-8 p: cppnpprp
6-7 k: jrnvhkkgkkb
12-14 h: hhhhhhhhhhhhhn
10-20 j: gjwjjjpjjljjjjjsjjjj
16-17 n: nnnnnnnnnnnnknnnnn
5-6 f: cdpppf
12-13 z: bvpzqzzlwzccfszt
1-9 k: wdkkvtrprltkkltz
9-16 x: xxxjxdxhtzjxgxsbx
17-18 j: jjjjjjjjjjjjjjjjlrxj
2-5 x: xkknmxxxn
3-9 f: tfffhfffffff
11-16 r: rrlgrrqmrrprcrrrvjr
9-12 m: nrlmgmxvgkmpmqvntx
4-7 r: hrrrzrrnm
3-15 g: stgjgbgggvdljcgdgfg
6-9 j: jjjzbpjjjjjj
2-3 p: ppjp
9-10 t: tttttttttt
10-12 w: wwwfwvwvnkwdwww
11-14 d: nvljddddftrndzx
6-7 r: rlrrrbqcrqr
2-4 c: xbdl
11-18 c: ccccccccccvcccccccc
5-6 j: gjjjjjj
3-7 w: xwhnqlhnpfrvlkqqrp
6-19 q: qqqfxdglhqqkqqqjnhdq
8-9 d: ddsdddqsv
14-16 g: gggggggggggggggt
3-5 b: bbbnh
11-14 t: mtttwttfttttltq
8-10 x: xxxcxxxfxq
8-11 l: lfgvnlzrfcllnxss
1-3 t: gtttttttwgtptt
1-20 j: jjgjlqjjcjcjjjsjgjlj
8-14 v: dqgmngvvqnswvvrrvc
2-8 m: gmpkmmsmmmmm
5-6 m: mmmmtm
6-7 z: zfzrxzxz
4-15 k: wmwkhhskgwzqpnk
2-6 d: hddddfzdsdtvg
9-10 g: gggsggggggrgggqg
11-12 m: mmmmmmmmmmmm
5-7 w: wwhqqhwwwwd
4-9 k: gtkrkkkdfk
1-4 g: kgtvg
7-11 r: rhdbzjrwrcr
9-18 f: fjffffffdffkfffffvf
3-5 t: ttttt
13-14 q: qqqqqqqqqqqqcqqq
1-8 n: vnnnnwnqlbln
3-5 s: ssmss
2-4 q: qqqbkvqxh
3-4 n: nshscq
10-18 s: sssssssssnsssssssgs
9-10 v: vvvvvglvzddvvjv
5-16 s: sspssssrssssslss
4-7 t: ttptxtt
2-4 r: rrrrrxdrr
8-9 b: xqvbbbbbh
3-4 b: bbbbc
1-13 v: qhvvvvlvvvvvdv
4-8 l: lhkllblwllll
3-4 q: qhjgq
8-9 p: rpbpppshdpp
13-18 k: kkkkkkkkkkkkkkkkkx
8-9 r: rrrrrrrrr
5-10 f: bjfwwqvjwffpnl
1-3 l: pbhl
4-6 g: mggfgv
1-4 t: tttttt
5-10 t: llqrtccxtttntxjcdczk
13-14 q: qqqqmqqqdqqfqgqqqq
6-10 w: wpwmwwrgwwk
13-16 q: qqshgkqkqvscqtqq
4-5 t: qttdt
2-12 h: fpbbxgpskzjpkvtr
10-12 z: zrszvzzzzrzkznpzw
1-4 z: tgzz
3-6 c: cdwccvgm
1-15 q: jqqxkqrqmtfqpvbpq
4-6 v: vjvlvvv
1-7 l: hjlllwj
6-13 s: lrspslfwsgjks
3-4 p: dpvk
4-8 h: hsjjhhff
1-18 k: pkkkkkkkkkkkkkkkkkk
6-8 n: nvcnxqnnrnnn
7-8 p: ppppppvzp
12-17 h: hhhhhzhhhhhkhhhhph
13-17 p: ppmpppnppppplpbvcppp
8-14 m: tmmcjvmmjgcfmmnj
1-3 t: tttth
3-4 x: zxxx
1-3 d: kpwhxpctcgdbdkb
5-6 z: knnlzzzzxjrghzb
9-10 w: wwmwwwwxgww
9-11 b: bqbbfbbwbbqbb
2-3 d: ddxp
10-12 g: gggvgsgtgggggbg
3-4 b: gbbb
8-20 c: jccqcxjcqfncfcbccgcc
4-9 z: zzzztzzzzzb
2-12 x: gmfmzzrxsqbx
10-12 x: xxxzbxxjxxxkxb
3-4 v: szpvbvlvr
2-5 r: grqqr
4-8 q: mqqnvqvqqq
6-10 v: vdbvnvvxvvvnvvvvv
4-10 d: dddtdddddjd
3-4 l: klmc
10-13 k: kkbkkkvkhkvkkkkkl
1-3 v: vvkz
13-14 h: pkgffgcszgsghbcdtpm
11-16 c: dgccccccctchxbfm
8-9 b: bbbbbbbbbtb
1-9 h: hrnhhlphhh
17-18 s: ssssssssssssssssst
5-10 h: hhhhhzhhhhhx
2-7 k: dlrncbkkwp
3-5 k: kfqkk
4-8 d: ddddddddd
10-15 m: mmmmrmmmlpmmmrgmkmx
2-4 g: gsgggkm
7-9 l: llsllllvrzlxlrgglk
2-9 f: fzfbvfkff
13-14 p: pxppppppppppvf
15-17 z: nbkxkjtszptcndzdl
3-6 z: zzxzzhz
12-13 t: tttttttttttttt
5-10 s: pssbfsbsssnssvsb
9-18 c: ccxthchscjjdccvcncl
3-5 f: fndfqffffff
2-4 m: clwvcdjmz
7-8 b: bbbpbjbb
2-3 r: rbxx
3-8 j: jjjjjjjjj
5-10 v: zrwnvbrvmm
4-6 h: hhghhhhh
6-8 f: fnffzfwfgffpfc
5-6 q: qvqqqqkqqq
3-5 v: vgwvrgqc
9-10 p: pppphbbpphppb
10-12 l: rdjrcxkgflll
9-14 h: qwwxnhfhnfhhbhhjr
9-10 d: ddddpddddd
4-13 w: wwwwwwwwwwwwwwww
8-16 x: xrxxxxxxxdlxxxxxxx
6-10 p: pplppkpzwb
5-7 j: wpjjjgjjvjqjrgj
4-13 w: cwwswwwwwjwwdvw
5-6 j: jjjsjjj
4-5 c: ccfccqdqctllbm
7-13 c: bcccccccccccccc
4-5 p: pgptw
1-4 p: qpwmp
3-5 f: fftfff
2-6 m: mmhbpmjlx
9-13 r: zrrczrlrzqxrszrrlrr
2-9 w: xrszwwwdkww
6-14 g: qlwswsgjdlgmzvwg
6-8 z: zzzzzzfz
6-7 w: xwzsjjr
2-13 f: fxmfdnfffkpggz
1-7 d: rdnlqdplfddjdd
11-14 j: kjjvjjjxpjjjjjzjjjz
12-14 p: gpppwvmqpfpptpfppkpc
6-11 k: kkkkkhkknkkkkkgk
2-6 m: swpgmf
1-3 x: sxzqnxxv
12-18 n: nnnnlnnnnnnnnnnjnxnn
3-4 p: pppq
8-9 v: vvvvvvvpv
6-9 l: ltsldwtlhll
11-14 x: xxxxlxxhdtxxxxxsxx
4-5 q: qsqvqvq
5-6 b: sbbbbt
10-13 l: lllllslllrllll
1-5 m: zmmmmmmm
2-10 s: ssssssssss
5-8 r: qqcsbjcrljvksc
11-15 n: bndcnknnxnnnnnbn
14-15 v: tvvvvvvvvhqwvvv
2-3 s: jsvcsb
12-15 d: mddddfddddlxddddd
12-18 w: wwznwcgwwwwwwhwwgwgw
2-9 v: dxlvvlvqg
2-5 h: njpsxr
11-13 m: rmmmmmmqmmzmtmm
4-8 m: smmmmmmm
3-10 j: ngwtkjfrjjtrxfnvj
2-3 p: ppwwpp
3-5 c: cbmcncnqc
11-13 m: mmmmmmmmmmcmm
4-14 g: gggggggggggggxg
2-4 b: bbfb
16-19 p: ppppppppcppdpppmppwp
10-12 g: gghgxgkvggrggrg
1-8 b: wjjqktjbbfkdz
8-11 c: cqccdcncfcccnc
2-10 z: qzvlqltzns
3-7 p: kgpmpzpbx
1-6 m: mxmdmlmmzmmmmmmmmm
1-7 r: rrrrrrrrrrrc
8-10 h: whhhqhhhjjhnch
6-10 x: xwxxxxkpxdkdnxvh
10-12 z: zzzzzzkzzpzzzzz
1-12 s: dssssssssssds
4-7 f: fvzbfcf
5-6 w: fwwxfwmswxkd
14-15 g: gggggggzgggggfhg
6-7 j: jjjjjnb
1-6 v: vcgksvnrbgsvkrwsc
6-15 w: wkvwwwpqwwwbtwxvhw
8-9 m: smmmmmmmmm
14-15 r: rrjrlrrrrrrrrzbr
3-8 w: tjqncxsnqkdpwzm
4-5 k: kkkkk
8-11 z: zzzzzzpzzxzzkzglzrz
5-6 k: kkkktzk
18-20 w: wwwscwwwwwwwwwwwwgws
1-12 z: dsfxzctdzzjxbt
14-15 q: qqqqqdqqqqqqqqq
7-10 r: rrxrrqrrxcbr
1-13 d: fmgjzgqhdhddd
4-6 g: mnggxttn
6-7 v: gwtmscl
11-14 g: gggnmgggggjggcgggrgg
17-18 k: gcslskskkkwkwkklkk
14-16 r: rrrrrrrrrrrrrbbrr
13-15 s: sjsxszssgfssdgwsbbsp
1-9 d: dddddddddddd
1-13 j: mnjjzckxzjjdj
1-2 j: hwrqjxxcj
6-7 z: fzzzwxnnrg
2-10 x: wxnqjddpxxpxw
3-10 s: srvvssdcvzfn
3-11 w: gwwwmwvwwww
3-4 q: qnqs
11-12 d: dddqdddzdrddddd
2-15 l: wldlltxhrwzzlhz
4-7 g: vgsgkfgmc
7-9 t: jdsltcgttdttd
9-10 q: xvqjqqqwqqqqqpqq
6-9 b: bbkbnbgbbb
2-5 t: ttcjtbkb
7-8 r: rdrrqrnrrrrrvp
2-4 f: pnfqfffvpfffffhbffv
10-12 z: kzzzzzzzzrzpzszz
4-5 m: tmmgxlm
3-4 k: kwhhk
1-11 h: hhhndhhfhvhchmmqhz
6-9 m: lmmmmmmmtmmm
1-5 d: dwdkd
5-6 d: dddddq
7-9 j: jjjvgvmjdjbjj
3-6 w: gwzhpjwkg
10-11 h: hhlmftlhkhhnh
8-20 h: hhgnchpthzhhhhqqwhcz
6-7 s: sssssgfs
3-11 g: jfgqgbvqppjn
5-6 d: ddqkmbd
14-15 h: hhhhhhhhhhhhhphhh
12-13 l: fwkwphdlwmfwg
3-4 z: pmbncwz
3-4 w: lgwv
3-4 s: sshs
2-7 z: zzflznzzzlzl
1-2 k: klbjvbwpkkptb
17-18 r: rrrrrrrrrrrrrrrrlrr
7-8 l: mlqgltml
9-14 l: llljlllxllllvl
5-13 b: bbbbtbbbbbbbfbbbg
7-10 d: cddlddgdqkld
4-6 q: kqqqqq
5-11 r: rrrvrrrsrjrr
7-10 t: tttgtltttbt
13-14 h: hhhhhqhhhhhhndh
3-4 s: dsss
6-12 w: jwwtxwwwkkwbw
7-8 v: cvvgnvvv
12-16 q: qqqqqqqqxvqqqrqqqq
6-11 z: zzkvzvzzzxzzzz
4-9 s: lcrvsssmfs
6-13 c: ccccpdcbcbcgdcccgdcc
7-9 q: qpqqqqqqhq
1-10 k: kwfkkxrkwhtmkwknkx
9-20 x: dxzrxxxsxxxjxxsxdczx
15-16 c: ckcccdccccccgccdcccn
2-7 s: sscxswskxhchxpfs
2-3 h: hhlh
2-6 p: ppppppp
6-8 r: ztmrsjqtrk
1-2 r: drsdvpmzdn
6-7 v: vqvvvgp
5-6 g: gggggg
3-7 d: tfkddmjzjgdxzlkd
16-17 b: bbwbbbbbbbbbbbwbt
2-11 b: hbrwqqhkcrbbvmx
5-11 f: ffffqfffffffff
2-5 n: nnwhn
8-11 j: jxjsjjmjkjkjhjx
1-3 h: hhhh
5-12 t: pttttszttwvlt
7-8 q: mqqqqqqqtkq
7-10 z: xzzzzzdzzzzzzzz
6-7 l: ljzqklhmm
3-5 k: bkkkrk
6-8 x: xxxxxlxcxxxbx
6-7 j: jjqjcpjfjj
8-9 l: cwmllhslwv
14-15 z: zzzzzzzzrzzzzcz
11-12 f: sfccmxxffsfg
5-6 q: qqqqqqqq
12-15 t: tttttttttttttttt
7-12 q: pqwsnhqfzdtqxd
4-5 b: fbcbb
5-7 t: tttstft
2-4 h: vjmr
4-6 r: rrfrprr
18-19 x: xxkxxxxxxxxxxxxxxxxx
2-4 k: qklkc
3-6 s: svjsrs
14-15 j: pjjjhjjjjjjdjgj
2-15 l: lhllllllllllllwlll
2-7 q: qvqqgsq
2-6 z: zzlzzzn
8-15 p: pppppppspppzppgp
3-4 l: kfzbl
9-12 w: hqjwwtdwwlwwmw
3-4 k: lghgg
3-5 k: khgzkk
2-10 r: rrsjlmrsrr
2-5 j: jjjjj
10-12 j: jfgzbdbjwhjjjf
1-4 f: ffff
9-10 j: jjjwjjwzgtcjzj
1-6 m: mmmmmq
5-6 h: hhwhph
3-6 w: wwpwwz
5-6 q: qqqkcb
8-9 f: ffffffxrff
13-14 m: lqvpbrhgmxmkgj
4-6 q: pqqxqk
7-11 b: mbbnsvbbjbbmb
17-18 q: qqqqqqqqqqhqqqqqqqq
3-12 w: qzcwxswwzvwzb
10-11 l: llllllllxxs
1-4 c: bvcch
11-13 s: svssqfvssshst
6-7 l: lllllrl
5-8 n: nhqkfnggnpkbhpgntlht
1-14 g: pgrdgjggggpggc
12-16 n: hlnhkxpnnsfctsnq
6-8 c: cccftqcfqc
1-5 f: fbfff
1-15 t: mttptftctttptttt
6-11 t: cxtttnnfttt
7-11 h: hsrfjswhhhhhhhhc
12-13 b: bbbbbbbbbbbnfbvb
12-13 x: lzvxjxxxcxxxq
12-13 p: ltwhhhqvdjptbsw
7-9 z: zzzzzztzn
4-5 g: ggqlgg
8-10 r: rbqxwzfhlzt
3-5 g: grhgk
14-19 z: zrzzbzztszpzvzbdfzm
7-11 d: tdjbcdpdddb
3-7 j: jpqjjjj
1-2 m: mcxklcmmws
7-8 r: drqlcrfqrgrrxrrg
1-5 g: kgngw
8-10 r: wrrbfrrrpntjrrrgrr
1-4 s: sstf
3-7 b: fnjnpmfjlwb
2-5 s: pwjkj
9-11 t: tbtbtrsrfzntwqq
13-16 m: mmmmmmmmmmmmmmmmm
13-18 b: bbbbbbbpqlbbbtbtpxb
9-11 w: wclwwwrwwwkhww
7-13 b: bbbbbbbwbbbbb
8-12 t: sxlwtttttnhn
4-7 q: kmdqxcqswqw
3-4 r: rjrr
4-6 m: mdmdmmmm
6-17 h: kjrkchhnfhqqcrmhhhdh
6-9 j: vjjgcjjjpdj
7-9 x: xxxxsxxxxx
8-16 j: jjjjjgjnjjjjjjjczd
6-8 t: tttttttd
7-17 s: ssmsspvsssspssssfssl
3-10 f: cgffsqxhdmrvdf
1-3 f: wfpf
11-12 k: kkkkkkkkkkfqk
12-15 s: cxghswzsbttmvpscr
3-4 n: nnkn
7-17 w: wwlwnxwxwwswwwwwg
3-10 h: hhvhhhrhhhfghhcn
10-11 l: lllxlllllllsl
5-7 d: ddbddtddv
5-15 g: ggtggqggggzggggwggv
2-10 r: mrpdmtrzsrngstsr
9-15 t: bsbptttttqwzmmtgbgm
8-10 q: qqqqqqqqqqq
6-9 h: hlrbcsnhh
12-17 f: ffkjfffffffnftfmf
11-12 s: ssssssssssss
9-15 k: hpkskfkkfkkkkbkk
8-18 b: bbbbbbdbbbbbbjbbjw
2-14 p: nppppkkppppppw
11-12 s: sxhtlrtnbgss
8-10 q: lnwqqqqqqsqhgqgbq
2-3 l: llll
14-17 d: gdddtlfmrdscgsqcmkm
5-8 w: tmmskwpq
16-17 b: cbbbpzbbrrbnkbbbbgj
14-17 b: bbbjbbbbbbbxbvbdszbb
9-10 q: qqqqqtqqqqr
5-6 t: ctgttztrtn
2-3 f: gcfgw
3-8 d: ddzddddsddddddd
15-16 g: ggggggjggggggggg
4-10 m: mcmmmmmmmmmvmmmm
11-15 p: ptppxjppppmkgpvdp
12-13 n: nnnnnnnfnnnrn
1-7 t: tnbwtttmtt
1-14 c: ndncvjbkcmcjplcp
1-2 f: fffh
14-17 t: llzkwltkmdntwndqs
4-7 w: bffzzwdwtkwwv
4-8 p: zpbpqpppxpb
8-14 n: fdnnznhnnlhmtn
7-14 l: vltrjnzhplkshlt
1-2 d: ddbhsjtv
2-10 d: dddddbzrdrdd
4-10 p: pkppshsfmp
10-11 g: gmgggzvzmgjgggzqxggd
2-4 j: fjvjjqf
5-19 h: cstkhhrctkxhhsfzpph
6-16 s: sssssssssssssssks
17-18 v: vvvvvvvvsvvvvvvvtv
2-5 j: zjjjjj
2-3 r: xrrrqlxl
17-19 c: cccrcccgcwcchccccdcc
7-14 l: lllllllllllllcsl
12-14 p: jlmhpjgdzjkpnbmnbk
17-18 n: nnnnnnnnnnnnnnnnqv
8-10 b: lqzrtbzbrb
8-11 x: xgxxfxxxxkxx
5-9 x: xxxsnxnxk
2-11 t: stxbxwssmgthttx
6-13 t: ktbktrtqcnlrgtvtwt
10-12 q: qqqqqqqqqrqq
14-15 q: znptzmgllrfrqrf
2-13 f: ffffffcffffff
9-10 h: hrbhnhhhhhhhh
8-14 k: lskskkkfzkkkkmkkm
1-3 g: kbmswhmghvwvnwxzzk
2-4 v: vttv
2-3 h: qhhmhghbh
6-13 t: ttgtttttktttdttt
7-9 v: vvrrvvvvvv
5-6 z: hcfwdr
5-13 g: sgmgggcggggggjrggggk
10-14 k: kzkkkkkkdskkktkkk
1-2 j: ptjj
4-11 b: xnbbbbnbbbb
14-16 b: bbbbbbbbbbbbbkbc
4-12 w: mwrwrcpwkpzwxrj
5-6 l: qlndzlcvw
1-2 s: ksxk
4-5 p: spgwql
7-10 j: ljjjjvxjhzjjjjrj
6-14 f: ffffffffffffff
11-15 d: ddddddddddzdpdn
3-5 t: ttmtz
6-7 x: kxvxxxx
3-6 m: mhmklm
4-8 h: bnhjhphgqxphpnthh
3-4 w: gwpq
10-13 t: ttvttjhttcttrtthtb
1-5 f: ffffg
6-14 v: rtjjvpbvnvjwlvktvvf
8-9 n: nnknnnxmnnn
9-10 v: vvsvvsvvqcvvpp
9-11 x: xxxxxxxxxwl
3-6 h: bwqxnrhkhhwhhh
4-8 l: lvllrlcf
1-2 m: gtrm
2-4 z: zzzz
6-7 j: jjjjjjj
15-16 g: sggggggggqgclgggmggw
7-9 v: dvmskjsczkvjcpsv
3-11 d: hmdrdfdqrddzdddjdd
7-9 m: mmmwfmmmtmm
6-8 x: lxsbxmlx
3-5 q: qqqxq
5-10 n: nnnnznnnnrn
4-6 m: zmmbzg
10-11 c: cpdndvlfltc
2-4 k: kzkkk
1-14 m: jmmmmmzmqtmsmmnmm
3-13 h: gwhhrlbstldmh
2-5 l: rrlhdqkzst
1-6 s: sqgsjs
3-4 h: hhgpb
11-16 l: llllllllllllllwkllll
7-11 j: gjjjtvhjjpj
14-15 h: hjjvmbhfjhhxkbh
1-2 t: rwxgz
13-18 p: pppppppppppppppppppp
1-14 x: xxrxxjxhxxxxxxbxxx
4-6 b: fbzsbpbbb
12-16 j: fjjgmnjwbvhjzzzs
6-7 m: mmmmmtq
12-16 n: nnmnnpnnnnnknqxkjnjn
5-8 w: wmwwwwpwlwllzwkcw
7-10 q: gbpsnqwqfm
10-13 x: vxxxxxxxxxdqnxx
7-12 q: qqqqqqqqqqqqqqq
1-2 n: snnn
6-10 l: lljllklllwp
3-4 b: bsnjzbb
13-14 g: tggztfgdggngmglgg
6-16 l: bkcwjlwcnfwthlll
8-11 s: sssssssvsss
3-4 k: xmthtrcx
7-8 h: wpmjhbgg
4-5 p: pcpkvp
14-15 s: ssrssssssssssjs
14-15 x: xxxzxxgxxtxxxxqx
6-7 m: nwmfmxmm
6-8 k: hsknkmvhkgkkfzkjf
4-5 s: sssms
7-9 g: gfgqldxgxdjzglcgg
10-11 k: ckkthkzpdrfv
1-14 p: jptppkcppjpppppppp
10-11 x: pxkccxpxdsq
2-8 x: xsgxxxxvgxxk
6-14 j: jjjjjzjjjjjjjtjjjj
7-10 x: rxxnxrzgxxd
6-12 g: dmgggpgggwczggghggm
3-6 h: hdhjhhhhchh
11-12 r: zrrkcrrrrrlh
7-9 v: vhqvlvwvzqwqvrxvjnf
1-5 r: rvmjr"""

In [ ]:
raw_input

In [ ]:
split_input = raw_input.split("\n")

In [ ]:
split_input

Convert
3-5 f: fgfff


to
{
min: 3
max: 5
char: "f"
password: "fgfff"
}

In [ ]:
def convert_to_dictionary(line):
    dictionary = {}
    line_as_list = line.split()
    
    min_max = line_as_list[0].split("-")
    dictionary["min"] = int(min_max[0])
    dictionary["max"] = int(min_max[1])
    
    dictionary["char"] = line_as_list[1][0]
    
    dictionary["password"] = line_as_list[2]
    
    return dictionary

In [ ]:
convert_to_dictionary("3-5 f: fgfff")

In [ ]:
passwords_and_policies = [convert_to_dictionary(item) for item in split_input]

In [ ]:
passwords_and_policies

In [ ]:
def is_in_compliance(dictionary):
    character_count = dictionary["password"].count(dictionary["char"])
    return dictionary["min"] <= character_count <= dictionary["max"]

In [ ]:
is_in_compliance({'min': 3, 'max': 5, 'char': 'f', 'password': 'fgfff'})

In [ ]:
valid_passwords = [item for item in passwords_and_policies if is_in_compliance(item)]

In [ ]:
valid_passwords

In [ ]:
len(valid_passwords)